packages/zero-server/src/adapters/drizzle.ts

```diff
@@ -0,0 +1,195 @@
+  PgDatabase,
+  PgQueryResultHKT,
+  PgTransaction,
+} from 'drizzle-orm/pg-core';
+import type {ExtractTablesWithRelations} from 'drizzle-orm/relations';
+import type {Schema} from '../../../zero-schema/src/builder/schema-builder.ts';
+import type {
+  DBConnection,
+  DBTransaction,
+  Row,
+} from '../../../zql/src/mutate/custom.ts';
+import {ZQLDatabase} from '../zql-database.ts';
+
+export type DrizzleDatabase<
+  TQueryResult extends PgQueryResultHKT = PgQueryResultHKT,
+  TClient = unknown,
+> = PgDatabase<TQueryResult, Record<string, unknown>> & {
+  $client: TClient;
+};
+
+/**
+ * Helper type for the wrapped transaction used by drizzle-orm.
+ *
+ * @remarks Use with `ServerTransaction` as `ServerTransaction<Schema, DrizzleTransaction<typeof drizzleDb>>`.
+ */
+export type DrizzleTransaction<
+  TDbOrSchema extends
+    | DrizzleDatabase<TQueryResult, TClient>
+    | Record<string, unknown>,
+  TQueryResult extends PgQueryResultHKT = PgQueryResultHKT,
+  TClient = unknown,
+  TSchema extends Record<string, unknown> = TDbOrSchema extends PgDatabase<
+    TQueryResult,
+    infer TSchema
+  >
+    ? TSchema
+    : TDbOrSchema,
+> = PgTransaction<TQueryResult, TSchema, ExtractTablesWithRelations<TSchema>>;
+
+export class DrizzleConnection<
+  TDrizzle extends DrizzleDatabase<TQueryResult, TClient> & {
+    $client: TClient;
+  },
+  TQueryResult extends PgQueryResultHKT = PgQueryResultHKT,
+  TClient = unknown,
+  TTransaction extends DrizzleTransaction<
+    TDrizzle,
+    TQueryResult,
+    TClient
+  > = DrizzleTransaction<TDrizzle, TQueryResult, TClient>,
+> implements DBConnection<TTransaction>
+{
+  readonly #drizzle: TDrizzle;
+
+  constructor(drizzle: TDrizzle) {
+    this.#drizzle = drizzle;
+  }
+
+  transaction<T>(
+    fn: (tx: DBTransaction<TTransaction>) => Promise<T>,
+  ): Promise<T> {
+    return this.#drizzle.transaction(drizzleTx =>
+      fn(
+        new DrizzleInternalTransaction(
+          drizzleTx,
+        ) as unknown as DBTransaction<TTransaction>,
+      ),
+    );
+  }
+}
+
+class DrizzleInternalTransaction<
+  TDrizzle extends DrizzleDatabase<TQueryResult, TClient> & {
+    $client: TClient;
+  },
+  TQueryResult extends PgQueryResultHKT = PgQueryResultHKT,
+  TClient = unknown,
+  TTransaction extends DrizzleTransaction<
+    TDrizzle,
+    TQueryResult,
+    TClient
+  > = DrizzleTransaction<TDrizzle, TQueryResult, TClient>,
+> implements DBTransaction<TTransaction>
+{
+  readonly wrappedTransaction: TTransaction;
+
+  constructor(drizzleTx: TTransaction) {
+    this.wrappedTransaction = drizzleTx;
+  }
+
+  async query(sql: string, params: unknown[]): Promise<Iterable<Row>> {
+    const stmt = fromDollarParams(sql, params);
+    const result = await this.wrappedTransaction.execute(stmt);
+    return toIterableRows(result);
+  }
+}
+
+/**
+ * Turn `$1, $2...` placeholders into a Drizzle SQL object with bound params.
+ */
+export function fromDollarParams(text: string, params: unknown[]): SQL {
```

Comment:
scary but seems reasonable. We're only ever sending keywords and identifiers so there _shouldn't_ ever be a $ coming in here except as a parameter.